kernel/aster-nix/src/syscall/mknod.rs

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use super::SyscallReturn;
+use crate::{
+    device::get_device,
+    fs::{
+        file_table::FileDesc,
+        fs_resolver::{FsPath, AT_FDCWD},
+        utils::{InodeMode, InodeType},
+    },
+    prelude::*,
+    syscall::{constants::MAX_FILENAME_LEN, stat::FileTypeFlags},
+    util::read_cstring_from_user,
+};
+
+pub fn sys_mknodat(
+    dirfd: FileDesc,
+    path_addr: Vaddr,
+    mode: u16,
+    dev: usize,
+) -> Result<SyscallReturn> {
+    let path = read_cstring_from_user(path_addr, MAX_FILENAME_LEN)?;
+    let current = current!();
+    let inode_mode = {
+        let mask_mode = mode & !current.umask().read().get();
+        InodeMode::from_bits_truncate(mask_mode)
+    };
+    let file_type = FileTypeFlags::from_bits_truncate(mode);
+    debug!(
+        "dirfd = {}, path = {:?}, inode_mode = {:?}, file_type = {:?}, dev = {}",
+        dirfd, path, inode_mode, file_type, dev
+    );
+
+    let (dir_dentry, name) = {
+        let path = path.to_string_lossy();
+        if path.is_empty() {
+            return_errno_with_message!(Errno::ENOENT, "path is empty");
+        }
+        let fs_path = FsPath::new(dirfd, path.as_ref())?;
+        current.fs().read().lookup_dir_and_base_name(&fs_path)?
+    };
+
+    if file_type.contains(FileTypeFlags::S_IFREG) || file_type.is_empty() {
+        let _ = dir_dentry.new_fs_child(name.trim_end_matches('/'), InodeType::File, inode_mode)?;
```

Comment:
Why is valid for the user to give a file path that ends with `/`, such as `/a/b/c/`, for creating a regular file?

Reply:
Thank you for pointing that out. Indeed, creating a regular file or a device with a path ending in `/` can cause issues. In Linux, using a path like `/a/b/c/` to create a file will return `File exists` if the directory exists or `No such file or directory` if it doesn't.

To address this, I've implemented the following function to parse the pathname up to the last `/`. With this addition, using mknod to create a regular file or a device with a path ending in `/` will return the same errors as in Linux.

```rust
/// Lookup the dir dentry until the last slash of the given path.
pub fn lookup_dir_until_last_slash(&self, path: &FsPath) -> Result<(Arc<Dentry>, String)> {
    let (dir_dentry, name) = self.lookup_dir_and_base_name_inner(path, false)?;
    if name.ends_with("/") {
        let dir_dentry = dir_dentry.lookup(name.trim_end_matches('/'))?;
        Ok((dir_dentry, String::from(".")))
    } else {
        Ok((dir_dentry, name))
    }
}
```
With this change, the code will now correctly handle paths ending in `/` for mknod, providing the appropriate error messages similar to Linux behavior.

Reply:
This new `lookup_dir_until_last_slash` does not seem to be in the same league as other `lookup` methods. It looks quite ad-hoc, designed to solve the special need of mknod. Its name (containing the word "slash") is ugly. In addition, the semantics of `openat` with `O_CREAT` flags and `mknodat` are quite similar. I am wondering why the former does not need this method. Given these signs of bad code, I encourage you to think twice: is it adding this method the best way to implement `mknodat`?

Reply:
Yes, you are right. This method is indeed very ugly and ad-hoc. However, I encountered many issues while writing `mknod`, and with the current methods, it was difficult to handle `mknod` properly. The main problems I found are as follows:

# Main Issues

## Method Definitions in the VFS Layer

There is a lot of redundant code in the current VFS layer's method definitions, which define operations on a case-by-case basis. For instance, given the pathname `/a/b/c`, there are two potential operations: one is to resolve the entire pathname, and the other is to leave the last component `c` for other processing. The VFS layer currently defines two separate methods for these cases:

```Rust
fn lookup_inner(&self, path: &FsPath, follow_tail_link: bool) -> Result<Arc<Dentry>> {
    let dentry = match path.inner {
        FsPathInner::Absolute(path) => {
            self.lookup_from_parent(&self.root, path.trim_start_matches('/'), follow_tail_link)?
        }
        FsPathInner::CwdRelative(path) => {
            self.lookup_from_parent(&self.cwd, path, follow_tail_link)?
        }
        FsPathInner::Cwd => self.cwd.clone(),
        FsPathInner::FdRelative(fd, path) => {
            let parent = self.lookup_from_fd(fd)?;
            self.lookup_from_parent(&parent, path, follow_tail_link)?
        }
        FsPathInner::Fd(fd) => self.lookup_from_fd(fd)?,
    };
    Ok(dentry)
}

fn lookup_dir_and_base_name_inner(
    &self,
    path: &FsPath,
    follow_tail_link: bool,
) -> Result<(Arc<Dentry>, String)> {
    let (mut dir_dentry, mut base_name) = match path.inner {
        FsPathInner::Absolute(path) => {
            let (dir, file_name) = split_path(path);
            (
                self.lookup_from_parent(&self.root, dir.trim_start_matches('/'), true)?,
                String::from(file_name),
            )
        }
        FsPathInner::CwdRelative(path) => {
            let (dir, file_name) = split_path(path);
            (
                self.lookup_from_parent(&self.cwd, dir, true)?,
                String::from(file_name),
            )
        }
        FsPathInner::FdRelative(fd, path) => {
            let (dir, file_name) = split_path(path);
            let parent = self.lookup_from_fd(fd)?;
            (
                self.lookup_from_parent(&parent, dir, true)?,
                String::from(file_name),
            )
        }
        _ => return_errno!(Errno::ENOENT),
    };
    // ...
    Ok((dir_dentry, base_name))
}
```

`lookup_inner` actually resolves the entire pathname, whereas `lookup_dir_and_base_name_inner` splits the pathname into two parts. For example, given `/a/b/c`, `lookup_dir_and_base_name_inner` splits it into `/a/b` and `c`, and then resolves `/a/b`. It is evident that both `lookup_inner` and `lookup_dir_and_base_name_inner` perform the same resolution of `/a/b`. However, separate functions have been defined for this. In fact, `lookup_inner` only needs to handle the last component `c` based on `lookup_dir_and_base_name_inner`.

## Issues of Code Redundancy in Syscall Open

The above example results in code redundancy, which is also evident in a segment of logic in the syscall `open`.

```Rust
let dentry = match self.lookup_inner(path, follow_tail_link) {
    Ok(dentry) => {
        let inode = dentry.inode();
        if inode.type_() == InodeType::SymLink
            && creation_flags.contains(CreationFlags::O_NOFOLLOW)
            && !status_flags.contains(StatusFlags::O_PATH)
        {
            return_errno_with_message!(Errno::ELOOP, "file is a symlink");
        }
        if creation_flags.contains(CreationFlags::O_CREAT)
            && creation_flags.contains(CreationFlags::O_EXCL)
        {
            return_errno_with_message!(Errno::EEXIST, "file exists");
        }
        if creation_flags.contains(CreationFlags::O_DIRECTORY)
            && inode.type_() != InodeType::Dir
        {
            return_errno_with_message!(
                Errno::ENOTDIR,
                "O_DIRECTORY is specified but file is not a directory"
            );
        }
        dentry
    }
    Err(e) if e.error() == Errno::ENOENT && creation_flags.contains(CreationFlags::O_CREAT) => {
        if creation_flags.contains(CreationFlags::O_DIRECTORY) {
            return_errno_with_message!(Errno::ENOTDIR, "cannot create directory");
        }
        let (dir_dentry, file_name) =
            self.lookup_dir_and_base_name_inner(path, follow_tail_link)?;
        if file_name.ends_with('/') {
            return_errno_with_message!(Errno::EISDIR, "path refers to a directory");
        }
        if !dir_dentry.mode()?.is_writable() {
            return_errno_with_message!(Errno::EACCES, "file cannot be created");
        }
        dir_dentry.new_fs_child(&file_name, InodeType::File, inode_mode)?
    }
    Err(e) => return Err(e),
};
```

Taking the pathname `/a/b/c` as an example, you can see that this logic first uses `lookup_inner` to search for the entire pathname. If it does not find the corresponding file, it then calls `lookup_dir_and_base_name_inner` to resolve `/a/b/`. It is evident that the segment `/a/b/` has been resolved twice in this logic. However, we only need to resolve it once, and then handle the last component `c` accordingly.

## Handling of the Last Component

The above syscall `open` needs to handle the last component `c` accordingly, and other syscalls also need to handle the last component differently based on different situations. They can be roughly categorized into the following types:

- Need to open the last component, such as `open`.
- Need to create a file based on the last component, such as `mkdir`, `mknod`, `link`, `symlink`.
- Need to perform other operations based on the last component, such as `rename`, `rmdir`, etc.
- Need to resolve the entire pathname directly, such as `stat`, `chmod`, etc.

This categorization is mainly based on the fact that each category has similar error handling during pathname resolution. However, the current system calls have very inconsistent error handling, which is also incompatible with Linux. For example, the syscall `link` does not return `EISDIR`. For these reasons, I believe it is necessary to define methods suitable for each category and ensure that error handling is compatible with Linux. Another example is when I was writing `mknod`, I found that the existing methods were not completely suitable for `mknod`, so I forcibly defined an ugly method `lookup_dir_until_last_slash`.

# Need to do

Based on the issues found above, I think we need to improve the methods in the VFS layer. I believe it is unnecessary to provide `lookup_XXX` methods to others. Based on the analysis above, only the handling of the last component differs, so I think the main actions needed are:

1. Only one `lookup_XXX` method is needed to resolve the pathname up to the last component, similar to the previous `lookup_dir_and_base_name_inner`. This method should be private.
2. Define methods corresponding to each category of syscalls, and these methods should be public. For example:
   1. For opening the last component, define `path_open()`. The main logic is to use `lookup_dir_and_base_name_inner` and then handle the opening of the last component accordingly.
   2. For creating a file based on the last component, define `path_create()`. The main logic is to use `lookup_dir_and_base_name_inner` and then handle the unified processing of the last component for creation operations, such as returning unified errors.
   3. For other operations based on the last component, such as `rename`, `rmdir`, etc., define `path_parentat()`. The main logic is to use `lookup_dir_and_base_name_inner`, and then let each syscall handle the last component, including error handling.
   4. For directly resolving the entire pathname, such as `stat`, `chmod`, etc., define `path_lookup`. The main logic is to use `lookup_dir_and_base_name_inner` and then resolve the last component.

The above is a general design. I will further summarize the VFS layer and propose an issue for specific naming and design.

# Handling of `mknod`

Based on the analysis above, I defined a `path_create` as a workaround and applied this method to `mkdir`, `link`, and `symlink`.

```Rust
/// The method used to create a new file using pathname.
///
/// For example, mkdir, mknod, link, and symlink all need to create
/// a new file and all need to perform unique processing on the last
/// component of the pathname. path_create is used to provide a unified
/// method for pathname lookup and error handling.
pub fn path_create(
    &self,
    path: &FsPath,
    lookup_flags: LookupFlags,
) -> Result<(Arc<Dentry>, String)> {
    let (dir_dentry, filename) = self.lookup_dir_and_base_name_inner(path, false)?;
    if dir_dentry.lookup(filename.trim_end_matches('/')).is_ok() {
        return_errno_with_message!(Errno

::EEXIST, "file exists");
    }

    if !lookup_flags.contains(LookupFlags::LOOKUP_DIRECTORY) && filename.ends_with('/') {
        return_errno_with_message!(Errno::ENOENT, "No such file or directory");
    }

    Ok((dir_dentry, filename))
}
```

I will integrate the introduction of other functions in subsequent updates.

Reply:
@liqinggd Could you add your comments to the mentioned issues? Thanks!

Reply:
Hi, the reason that we define `lookup_inner` and `lookup_dir_and_base_name_inner` is that the lookup method should handle the case where the `FsPath` is just a `cwd` or a `fd`. In this case, you can not lookup the parent dir.

Reply:
BTW, the `open` method is not efficient when we should create a new inode, so we should improve it.

Reply:
> Hi, the reason that we define `lookup_inner` and `lookup_dir_and_base_name_inner` is that the `lookup_inner` method should handle the case where the `FsPath` is just a `cwd` or a `fd`. In this case, you can not lookup the parent dir. But in the case of `lookup_dir_and_base_name_inner`, the common case is used to some syscalls like `link`, `mkdir` or `rename`, This syscalls must ensures that the path is not a `cwd` or `fd`.

I think what you said corresponds to the situations I mentioned.
> Need to open the last component, such as open.
Need to create a file based on the last component, such as mkdir, mknod, link, symlink.
Need to perform other operations based on the last component, such as rename, rmdir, etc.
Need to resolve the entire pathname directly, such as stat, chmod, etc.

`lookup_inner` and `lookup_dir_and_base_name_inner` simply divide these cases into two types. And although `lookup_inner` handles cwd and fd, these two cases are not designed for pathname lookup, that is, the pathname lookup process still has some duplication.

Although system calls such as `link`, `mkdir` or `rename` have similar parts in pathname lookup, because they all leave the last component for special processing, it is not feasible to simply classify them into one category. As my classification shows, the processing of the last component can be further classified.

Below is the original symlink code:
```rust
if linkpath.is_empty() {
return_errno_with_message!(Errno::ENOENT, "linkpath is empty");
}
if linkpath.ends_with('/') {
return_errno_with_message!(Errno::EISDIR, "linkpath is dir");
}
let fs_path = FsPath::new(dirfd, linkpath.as_ref())?;
current.fs().read().lookup_dir_and_base_name(&fs_path)?
```
You can see that if the path name ends with /, EISDIR will be returned, but in fact, according to the [Linux manual](https://man7.org/linux/man-pages/man2/symlink.2.html), `symlink` does not return `EISDIR`. In fact, `mkdir`, `mknod`, `link`, and `symlink` handle the last component of the path name in a similar way, including the return of errors. Now, the processing of the last component by these system calls, including error returns, needs to be completed by each system call itself, without a unified definition, which leads to different error returns.

And some other system calls can also classify the processing of the last component, such as the need for `lookup_inner`.

Therefore, I think in order to reduce code redundancy and adapt to various situations more comprehensively, we need to redesign the methods provided by VFS.

Reply:
> BTW, the `open` method is not efficient when we should create a new inode, so we should improve it.

I think this problem can be solved after we redesign the approach of VFS.

Reply:
@Plucky923 You have done an excellent job in summarizing the current status. So you plan to refactor existing method of `VfsResolver` in another PR, right?

Reply:
> @Plucky923 You have done an excellent job in summarizing the current status. So you plan to refactor existing method of `VfsResolver` in another PR, right?

Yes, but the classification of VFS is not detailed enough at present. I will raise an issue after a more detailed summary, and then raise a PR for comprehensive revision.

Reply:
@Plucky923 Yes, currently, many of our system calls still have issues with returning error codes. Your summary of how different syscalls handle the last component is excellent. I'm looking forward to your new PR to address this problem.